### ISSUE TITLE
Unable to resolve module

### ISSUE THREAD

**Issue by tedjames** [facebook/react-native#11498]
### Description
Unfortunately, I’ve run into a devastating bug that prevents me from loading any new modules into my app. I first discovered this bug when reaching the part of a react native course by Stephen Grider where he goes into networking with Axios. 

After installing Axios with NPM, I imported it, saved index.ios.js, and refreshed the simulator. After refreshing I was greeted with the message shown below:

<img width="1189" alt="bug3" src="https://cloud.githubusercontent.com/assets/13091675/21242561/f16694c6-c2e2-11e6-8f35-a03ea7c493ea.png">

Whenever I removed the import statement for Axios and refreshed, the error would go away:

<img width="1203" alt="bug4" src="https://cloud.githubusercontent.com/assets/13091675/21242579/07cb2f24-c2e3-11e6-9f7c-c15233e7ea08.png">

### Reproduction
To reproduce this, I'd suggest building the app with NPM and/or Yarn running on Node 4.7 or the latest LTS release, Node 6.9.2

I can provide a repository later today if that would help...

### Solution - FOUND
Just solved the issue! Here's the solution:
Try this:
1. Click "Reset content and settings..." in the simulator's drop down menu and confirm
2. Close the simulator after it reloads
3. Close the terminal window running the watchman
4. Run the following commands within terminal inside of your project's directory:
```
watchman watch-del-all
rm -rf ./node_modules
npm cache clean
yarn cache clean
rm -rf $TMPDIR/react-*
yarn install
npm cache clean
yarn cache clean
yarn add uuid
```
5. Rebuild the project inside of Xcode and be sure to make any suggested modifications prior to running the app.

### Additional Information

* NPM version: 2.15.9 -> 4.0.3 (just updated to the latest version and the bug remains)
* Yarn version: 0.17.10
* Xcode version: 8.2 (8C38)
* React Native version: 0.39.2
* Platform: iOS
* Operating System: MacOS Sierra 10.12.2

I was really excited to finally get started with React Native but this bug absolutely kills my ability to proceed with any tutorials/courses. I'd really appreciate if someone could help me out with this.

Thanks :)

**Comment by tedjames** [comment]
Just solved the issue! Here's the solution:

1. Click "Reset content and settings..." in the simulator's drop down menu and confirm
2. Close the simulator after it reloads
3. Close the terminal window running the watchman
4. Run the following block of code inside of your project's directory:
```
watchman watch-del-all
rm -rf ./node_modules
npm cache clean
yarn cache clean
rm -rf $TMPDIR/react-*
yarn install
npm cache clean
yarn cache clean
yarn add uuid
```
5. Rebuild the project inside of Xcode and be sure to make any suggested modifications prior to running the app.

**Comment by lacker** [comment]
I'm glad you solved it - thanks for also posting the solution here to help other folks!

**Comment by tedjames** [comment]
Hey @lacker, I'm still experiencing this issue. Unfortunately, the solution I provided above is more of a workaround. Anytime I install a package with npm or yarn on any project, I have to go through the steps above and more often than not, also restart my Macbook.

**Comment by ouzhou** [comment]
@tedjames I have encountered the same problem

**Comment by ThunderBeard** [comment]
@tedjames I have the same issue and thank you for posting a solution/workaround. But being the N00b I am I need some more instructions to how I run the commands you suggest. I try to use them in the CMD in my project folder, but get the response that commands are not recognized... Could someone elaborate the instruction please?

Thank you in advance

**Comment by ouzhou** [comment]
@ThunderBeard, rm -rf is not work in windows

**Comment by ThunderBeard** [comment]
Thanks, so are there any instructions for windows then?

**Comment by ouzhou** [comment]
I find a better way to solve your problem, you can use git bash to run those commands
![image](https://cloud.githubusercontent.com/assets/7693866/21749488/232e3ad6-d5da-11e6-9100-cba493b839e2.png)

**Comment by tedjames** [comment]
Hey, @ThunderBeard and @ouzhou, what operating systems are you guys on? I have a feeling that this issue can be reproduced on Windows as well: #11772

**Comment by tedjames** [comment]
Found another case similar this running on MacOS: #11257

**Comment by jussisaurio** [comment]
This happens to me too now. Problem started with Axios for me as well.

**Comment by anthony-skr** [comment]
This happens to me too with Axios / RN 0.40.

**Comment by agarcia17** [comment]
same thing here. on Windows - Axios / RN 0.40
But the same project runs fine on my Mac (both Android emulator and iOS simulator)

**Comment by boyney123** [comment]
Nothing seems to work for me, trying to install about 10 different XML parsers in R.N and just getting this error message.

Anybody got any more solutions, almost tried everything :/

**Comment by GantMan** [comment]
### Solved but in a terrible way
TL:DR; use this in your package.json for now
```
    "axios": "infinitered/axios#windows_friendly",
```

For some reason packager traces down a require inside a conditional, but only on Windows machines!  **How?** I have no clue.  Perhaps @cpojer / @ide _et al._ can illuminate how this could be?  [This require](https://github.com/mzabriskie/axios/compare/master...infinitered:windows_friendly) should never be called, but somehow it is traversed by packager.

![image](https://cloud.githubusercontent.com/assets/997157/22135070/7fdf8388-de91-11e6-8212-d55b154dc84f.png)

Using my branch in npm (which comments out the offending require) fixes everything.

### This issue in other tickets
* Axios repo is tracking the issue here:  https://github.com/mzabriskie/axios/issues/648
* Another react-native ticket has been opened here:  #11887 
* For those of you having this problem in apisauce: tracked here (https://github.com/skellock/apisauce/issues/47), you can get around this by using this branch, that uses the aforementioned branch.

```
    "apisauce": "infinitered/apisauce#windows_friendly",
```

I wouldn't mind hearing the solution to this.  I'm sure @skellock (who helped me track this down) wouldn't mind seeing the solution either.

**Comment by tedjames** [comment]
Problems just went away for me recently and for no apparent reason. Not too sure what I've done different besides the fact that I've tried RN 0.38 thru 0.40. 

Now I'm trying to wrap my head around some of these breaking changes introduced with RN  0.40 -__-

**Comment by rubennorte** [comment]
Hi, I'm an axios contributor. This conditional `require` works because when the compilation target of the bundle is "web" the "http" adapter is replaced with the "xhr" one. See https://github.com/mzabriskie/axios/blob/68ec2abc4a74777f746824b5595d58e07b44c119/package.json#L72

Has there been any change in the way RN resolves that?

**Comment by skellock** [comment]
Yes.  In 0.37 there was reorganization of these files (maybe when the bug arrived?).

Recently [this PR](https://github.com/facebook/react-native/pull/11641/files) fixing `path.sep` issues by @jhen0409 was accepted and merged into [0.41.0-rc-0](https://github.com/facebook/react-native/releases/tag/v0.41.0-rc.0).

Looks like we're good when 0.41 ships?  I can confirm this weekend, but hopefully someone beats me to it.

Thanks all!  I learned a lot here.

**Comment by Thibaut-Fatus** [comment]
We still have problems resolving new libraries on 0.41, manually installing the dependency or  resetting content and settings in the simulator sometimes helps, but it seems it is not yet fixed

**Comment by nhuthuynh** [comment]
i encounter the same problem but the workaround does not fix it. any other suggestion ?

**Comment by nhuthuynh** [comment]
i fixed the problem by following the workaround then install locally react-native, react and react-native-animatable which is the package cannot be resolved. Hope this help !

**Comment by whoadave** [comment]
@tedjames why the `yarn add uuid` in your fix?

**Comment by MehdiAlouafi** [comment]
Hi everyone, maybe it can help, I was having this issue because I was using a package that uses some nodejs utilities which are available on the global object.
In other word, this package could only work in a nodejs env like and not within' the react-native env. It was my mistake. 

Hope it helps

**Comment by sw-tt-vipulparmar** [comment]
move your project to another path

**Comment by PrimeLens** [comment]
This is happening with me on a mac.  Have reduced my project to a simple hello world that imports events from 'events' and error still occurs, the work around in the original post is not working even with a restart of the machine.
npm 4.2.0   and   node v7.10.0
react-native-cli: 2.0.1  and   react-native: 0.45.1
MacOSSierra 10.12.4   and   xcode  8.3.2

**Comment by tedjames** [comment]
@whoadave The yarn add uuid was unnecessary in hindsight. Just noticed a warning in my console asking to use uuid instead of node-uuid.

@PrimeLens Try wiping the cache of everything you listed, then uninstall and reinstall everything to their latest versions if possible. This includes xcode, rn-cli, rn, and node. Then start your project from scratch in a completely new directory. Be sure to set privileges to everyone on all directories and sub-directories within your project.

**Comment by PrimeLens** [comment]
@tedjames  good lord that solution makes this feel like the dark ages!!!  Ii'll try as you suggest, thanks for your response

**Comment by PrimeLens** [comment]
this has nothing to do with axios, same bug though

**Comment by moonvader** [comment]
Same bug :( 
I tried to use React Native and as I see there are many people like me that can't run simple project. If it is so hard to start - what will be in future?

**Comment by hellomarcoliver** [comment]
Still, this does not work for me. Any help is much appreciated. Check out my repository here: [https://github.com/hellomarcoliver/ReactNativeUIs/tree/master/Jobs](url)

Error:
![simulator screen shot - iphone 6s - 2017-10-06 at 11 40 44](https://user-images.githubusercontent.com/3780027/31275405-fed64022-aa8e-11e7-90a4-90eb15676fca.png)

**Comment by epenabella** [comment]
I had a similar issue.  Maybe mine seems stupid, but the guide I was using was telling me to create an artifacts folder for the compiled .js files.  But after I stopped the output of compilation going elsewhere, like the artifacts folder, and instead, just left the output of compiled .js files next to the .tsx files this problem went away instantly.

In other words, have your compiled .js files next to your .tsx files and this might fix your problem :)

**Comment by 11reed** [comment]
This is so frustrating & annoying, the only thing that works for me is to create a completely new project & move my files over. Are there any updates on this?

**Comment by AakashMaheshwari** [comment]
Hi there.. I am facing the same issue but this is not because of axios. This problem occurs when i import a custom component I made. For example, i made a component of AlbumDetails which when I imported in another component named AlbumList, I faced the issue. (Please refer to the screen shots below :)

Screenshot 1 is when the import statement is commented and so I don't get any error and Screenshot 2 is when the import statement is not commented I get the following errors
**ScreenShot 1**
![screen shot 2018-03-06 at 8 21 10 pm](https://user-images.githubusercontent.com/34339542/37039004-6260ee58-217c-11e8-992e-e8cc53ef4c03.png)

**Screenshot 2**
![screen shot 2018-03-06 at 8 22 44 pm](https://user-images.githubusercontent.com/34339542/37039061-851c9488-217c-11e8-861b-a34ee4f38748.png)

Please help me resolve this.. Thanks
